Add peer awareness design document

hackerwins opened the pull request:
<!--  Thanks for sending a pull request! -->

**What this PR does / why we need it**:

Writing a design document for Peer Awareness.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #48

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note

```

**Additional documentation**:

<!--
This section can be blank if this pull request does not require a release note.

Please use the following format for linking documentation:
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

**Checklist**:
- [x] Added relevant tests or not required
- [x] Didn't break anything

dc7303 commented on design/peer-awareness.md:
```diff
@@ -0,0 +1,64 @@
+# Peer Awareness
+
+Target version: Yorkie 0.1.2
```
I am not sure if it is absolutely necessary to specify the version. Can you explain why you entered the version?

hackerwins replied:
I thought writing a release version would allow us to think about the scope of features and focus more on our goals when writing your design documentation.

dc7303 commented on design/peer-awareness.md:
```diff
@@ -0,0 +1,64 @@
+
+Target version: Yorkie 0.1.2
+
+## Use case
```
What about `Overview` or `Goals` instead of `Use case`? I think that this section is more of an explanation of what problems `peer awareness` solves, rather than talking about use cases. What do you think?

hackerwins replied:
I think the goals could be suitable as well. But I wanted to write what value it would give the user with Peer Awareness.

habibrosyad replied:
Agree with @dc7303. Additionally I thought `Motivation` would suit this section as well, since this section mentions what's driving this feature.

dc7303 reviewed (commented):
This is really cool!
It seems to be well written for easy understanding. I gave some opinions, so please check.

habibrosyad reviewed (commented):
In my opinion, this design document is great, simple and easy to understand.

For design document I think we can introduce more solid and consistent structures (especially later when some collaborator/contributor want to convey their ideas for enhancing yorkie with more features). I thought a good reference for this would be the KEP (Kubernetes Enchancement Proposal), e.g. [this one](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1281-network-proxy).

hackerwins commented:
@dc7303, @habibrosyad Thank you for your review. I adjusted the section to reflect your comments.

Also, @dc7303 commented that the first document will be referenced in future documents, so it's a good idea to handle it carefully. So I added a template too.

dc7303 commented on design/peer-awareness.md:
````diff
@@ -0,0 +1,95 @@
+
+### How does it work?
+
+```
+ +--Client "A"----+                      +--Agent-----------------+
+ | +--Metadata--+ |                      | +--PubSub -----------+ |
+ | | { num: 1 } | <-- WatchDocuments --> | | {                  | |
+ | +------------+ |                      | |   docA: {          | |
+ +----------------+                      | |     A: { num: 1 }, | |
+                                         | |     B: { num: 2 }  | |
+ +--Client "B"----+                      | |   },               | |
+ | +--Metadata--+ |                      | |   ...              | |
+ | | { num: 2 } | <-- WatchDocuments --> | | }                  | |
+ | +------------+ |                      | +--------------------+ |
+ +----------------+                      +------------------------+
+```
````
I think we need to turn a diagram written in text like this into an image file. But I don't think it is now. When the branding of our project is done, it looks like we can draw it by applying our colors, designs, and feelings.
This can be a cumbersome task, but it seems that even if there are more documents, you can have unity. What do you think?

hackerwins replied:
Thanks for the comment. In my opinion, there seems to be no need to unify the style of diagrams in internal documents. However, it would be nice to apply it to the documentation we write on https://yorkie.dev.

dc7303 reviewed (approved):
LGTM 👍 I think the written template and the `peer-awareness.md` document are appropriate as our first design document.

mojosoeun reviewed (approved):
Thank you for your work! The design document and template are awesome and easy to understand! Also, I was able to learn about `peer-awareness`!
